### ISSUE TITLE
Partial<T> isn't strict enough

### ISSUE THREAD

**Issue by benneq** [Microsoft/TypeScript#29701]
**TypeScript Version:**  3.3.1

**Search Terms:** partial strict undefined union

**Code**

```ts
interface MyType {
    foo: string
}
  
const merge1 = (value1: MyType, value2: Partial<MyType>): MyType => {
    return {...value1, ...value2};
}

const merge2 = (value1: MyType, value2: Partial<MyType>): MyType => {
    return Object.assign({}, value1, value2);
}

const result1: MyType = merge1({ foo: 'foo' }, { foo: undefined });
// result1.foo is undefined AND of type MyType
const result2: MyType = merge2({ foo: 'foo' }, { foo: undefined });
// result2.foo is undefined AND of type MyType
```

**Expected behavior:**
Either `Partial<T>` should throw an error, because `foo: undefined` isn't allowed in `MyType`.
Or spread operator / Object.assign / union operator should throw an error.
Or merge1 / merge2 should throw an error, because return types don't match.

**Actual behavior:**
It's valid code.

**Playground Link:** 
http://www.typescriptlang.org/play/#src=interface%20MyType%20%7B%0D%0A%20%20%20%20foo%3A%20string%0D%0A%7D%0D%0A%20%20%0D%0Aconst%20merge1%20%3D%20(value1%3A%20MyType%2C%20value2%3A%20Partial%3CMyType%3E)%3A%20MyType%20%3D%3E%20%7B%0D%0A%20%20%20%20return%20%7B...value1%2C%20...value2%7D%3B%0D%0A%7D%0D%0A%0D%0Aconst%20merge2%20%3D%20(value1%3A%20MyType%2C%20value2%3A%20Partial%3CMyType%3E)%3A%20MyType%20%3D%3E%20%7B%0D%0A%20%20%20%20return%20Object.assign(%7B%7D%2C%20value1%2C%20value2)%3B%0D%0A%7D%0D%0A%0D%0Aconst%20result1%3A%20MyType%20%3D%20merge1(%7B%20foo%3A%20'foo'%20%7D%2C%20%7B%20foo%3A%20undefined%20%7D)%3B%0D%0Aconst%20result2%3A%20MyType%20%3D%20merge2(%7B%20foo%3A%20'foo'%20%7D%2C%20%7B%20foo%3A%20undefined%20%7D)%3B

**Related Issues:** 
none?

**Comment by benneq** [comment]
Seems to be a duplicate of https://github.com/Microsoft/TypeScript/issues/13195
